### PR TITLE
feat: use MkdirAll by OS(#19301)

### PR DESCRIPTION
**Pull request by thisishwan2**
Hello. This is a PR for issue https://github.com/argoproj/argo-cd/issues/19301

I created an implementation to run MkdirAll by os. And I abstracted it. In hydratorhelper, I modified it to use an implementation that fits os.
Please let me know if there is anything wrong!

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: commitserver/commit/mkdir_provider.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package commit
+
+import (
+	"os"
+)
+
+// MkdirAllProvider is an interface for creating directories.
+type MkdirAllProvider interface {
+	MkdirAll(path string, perm os.FileMode) error
```
Comment (crenshaw-dev) [inline commitserver/commit/mkdir_provider.go:9]: I think the interface we need is `func MkdirAll(root, unsafePath string, mode int) error`. The Linux implementation will just pass through the args to `securejoin.MkdirAll`. The non-Linux implementation will run `securejoin.SecureJoin(root, unsafePath)` and then run `os.MkdirAll`.

Comment (thisishwan2) [reply commitserver/commit/mkdir_provider.go:9]: Then, is it correct that you want to treat the part you uploaded on the issue with func MkdirAll (root, unsafe path string, mode int) error??
 
before

``` go
fullHydratePath, err = securejoin.SecureJoin(h.dirPath, hydratePath) 
 if err != nil { 
 	return fmt.Errorf("failed to construct hydrate path: %w", err) 
 } 
 // TODO: consider switching to securejoin.MkdirAll: https://github.com/cyphar/filepath-securejoin?tab=readme-ov-file#mkdirall 
 err = os.MkdirAll(fullHydratePath, os.ModePerm) 
```

after
``` go
var fullHydratePath string
err = MkdirAll(rootPath, hydratePath, os.ModePerm)
```

**Comment by thisishwan2** [comment]
@crenshaw-dev As you said, I modified the implementation.
`securejoin.MkdirAll`(in linux) and `securejoin.securejoin (root, issuePath)` and `os.MkdirAll`(in default) are implemented in their respective implementations
To use the fullHydratePath variable in future codes, 
`fullHydratePath, err = mkdirAllProvider.MkdirAll (rootPath, hydratePath, os.ModePerm)` Modified hydratorhelper.go code in form.

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
This looks nearly perfect! A few requests:

1) Can we eliminate the structs and just have a plain function called `SecureMkdirAll`? I tried locally, looks like the struct isn't necessary.
2) Can you add super basic unit tests for both implementations?

**Comment by thisishwan2** [comment]
Great! 
I think it's unnecessary to erase the interface `mkdir_provider.go` while erasing the structure, so I'll erase it together.

**Review by crenshaw-dev — APPROVED** [review]
Thank you!!

**Pull request merged**